# MapScapingMap

My plan is to create a map with the locations of all the interview partners Danial O'Donohue had on his [MapScaping](https://mapscaping.com/) Podcast.
For this, I scrape his website for the basic info (episode, date, duration, title, categories) on each of the >200 podcast episodes. Then, I have to add the interviewees and their location etc. manually.

### import packages

In [9]:
# import all necessary modules and packages
import requests
from bs4 import BeautifulSoup
from tqdm import tqdm
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
import time
warnings.simplefilter(action='ignore', category=FutureWarning)

In [11]:
pd.set_option('display.max_rows', None)  # Optionally, ensure all rows are shown

### scrape website

In [13]:
base_url = 'https://mapscaping.com/podcasts/page/'
page_numbers = range(1, 13)

# Creating a list of URLs for pages 1 to 12 /!\ except page 4, see below!
links_list = [base_url + str(page_number) for page_number in page_numbers if page_number != 4]
# links_list

In [15]:
all_dfs = []

for url in tqdm(links_list): # Iterate over the links and track progress
    request = requests.get(url)
    soup = BeautifulSoup(request.text)

    date_soup = soup.findAll('span', class_= "blog-meta-date-display")
    dates = [date.text for date in date_soup]

    title_soup = soup.findAll('h2', class_= "secondline-blog-title")
    titles = [title.text for title in title_soup]
    titles = [title.replace('\n','').replace('\t','') for title in titles] # not nice, but works

    episode_soup = soup.findAll('div', class_= "blog-meta-serie-episode")
    episodes = [episode.text for episode in episode_soup]

    categories_soup = soup.findAll('span', class_= "blog-meta-category-list")
    categories = [category.text for category in categories_soup]
    categories = [category.replace('\n','').replace('\t','') for category in categories] # not nice, but works

    duration_soup = soup.findAll('span', class_= "blog-meta-time-slt")
    durations = [duration.text for duration in duration_soup]

    current_df = pd.DataFrame({'Episode': episodes, 'Date': dates, 'Title': titles, 'Duration': durations, 'Categories': categories})

    # Append the current DataFrame to the list
    all_dfs.append(current_df)

    # Adding a 5-second delay
    time.sleep(5)

# Concatenate all DataFrames in the list
result_df = pd.concat(all_dfs, ignore_index=True)


100%|████████████████████████████████████████████████████████████████████████████████| 11/11 [01:16<00:00,  6.98s/it]


In [19]:
result_df['Date'] = pd.to_datetime(result_df['Date'])
result_df['Episode'] = result_df['Episode'].str.replace('Episode ', '')
result_df

Episode       Date                                              Title  \
0       237 2024-09-25                        Female Voices in Geospatial   
1       236 2024-09-18                                             Qfield   
2       235 2024-08-28                                Analyst To Engineer   
3       234 2024-08-16                        Satclip – Encoding Location   
4       233 2024-08-01                         Natural Language Geocoding   
5       232 2024-07-10                     Semantic Search For Geospatial   
6       231 2024-06-05                   Why You Should Care About L Band   
7       230 2024-05-23                           GeoParquet for beginners   
8       229 2024-05-16                              Finding Stuff Indoors   
9       228 2024-05-01                          What is humanitarian GIS?   
10      227 2024-04-12                           AI Autocomplete for QGIS   
11      226 2024-03-21  GNSS receivers – why precise positioning will ...   
12      225 2024-03-15  The Way You Talk About Your Geospatial Skills ...   
13      224 2024-02-29                                  Modern Geospatial   
14      223 2024-02-15               Introduction To LIDAR & Point Clouds   
15      222 2024-01-26            Introduction to Cloud Native Geospatial   
16      221 2024-01-09                                             GeeMap   
17      220 2023-12-27                                     GPS Reimagined   
18      219 2023-12-20                   The Business of QGIS Development   
19      218 2023-12-14                      Making Beautiful Maps In Felt   
20      217 2023-12-06    Planet – Imaging Everything, Every Day … Almost   
21      216 2023-11-30  Fire Mapping, Maritime Search And Wide Angle I...   
22      215 2023-11-22                    Personal Branding in Geospatial   
23      214 2023-11-15                    Entity Resolution with Placekey   
24      213 2023-10-04                        Strategic Buy-In For FOSS4G   
25      212 2023-09-19                                   From GNSS to VPS   
26      211 2023-08-30        Overture Maps And The Daylight Distribution   
27      210 2023-08-16                       100 billion Points Every Day   
28      209 2023-08-09                   Synthetic Data For Real Problems   
29      208 2023-08-03                                          Hub Ocean   
30      207 2023-07-19                             Felt – Upload Anything   
31      206 2023-07-05                                   The Rapid Editor   
32      205 2023-06-28  PostgreSQL – Listen and Notify Clients In Real...   
33      204 2023-06-22  Applying For A Job, Getting Picked and Negotia...   
34      203 2023-06-14                 Using Lasers To Talk To Satellites   
35      202 2023-06-07    From Pixels to Patterns: AI in Spatial Analysis   
36      201 2023-05-31              pygeoapi – A Python Geospatial Server   
37      200 2023-05-24                            Big Data In The Browser   
38      199 2023-05-17                             Rasters In A Database?   
39      198 2023-05-10                  ChatGPT and Large Language Models   
40      197 2023-05-07                           Spatial Knowledge Graphs   
41      196 2023-04-26                          Computer Vision and GeoAI   
42      195 2023-04-19                     Designing for Location Privacy   
43      194 2023-04-13                     Hyperspectral vs Multispectral   
44      193 2023-04-02                     All Of The Places In The World   
45      192 2023-03-22           Planet Scale Tiled Maps Without A Server   
46      191 2023-03-15                    Storytelling With  Point Clouds   
47      190 2023-03-08                             Geospatial Archaeology   
48      189 2023-03-01       Navigating the World of Geospatial Standards   
49      188 2023-02-22               Making Money With Geospatial Content   
50      187 2023-02-15                       Distributing Geospatial Data   


On page 4 of the website:


Episode 162 (QGIS offline and in the field, July 20th, 2022) on page 4 does not have info on duration and episode, which makes it hard to make a dataframe out of it (as the arrays have different length)
In the code below, I manually insert the missing information and bind page 4 to the rest of the scraped data. 

Episode 162 is located:
between 'whitebox tools' and 'sentinel hub'
between 161 and 163

In [23]:
base_url = 'https://mapscaping.com/podcasts/page/4'

response = requests.get(base_url)
soup = BeautifulSoup(response.text)

In [24]:
date_soup = soup.findAll('span', class_= "blog-meta-date-display")
dates = [date.text for date in date_soup]
print(len(dates))
dates

21


['September 28, 2022',
 'September 22, 2022',
 'September 21, 2022',
 'September 14, 2022',
 'September 8, 2022',
 'August 31, 2022',
 'August 18, 2022',
 'August 10, 2022',
 'August 8, 2022',
 'August 3, 2022',
 'July 27, 2022',
 'July 20, 2022',
 'July 20, 2022',
 'July 6, 2022',
 'June 29, 2022',
 'June 25, 2022',
 'June 8, 2022',
 'June 1, 2022',
 'April 28, 2022',
 'April 20, 2022',
 'April 13, 2022']

In [27]:
title_soup = soup.findAll('h2', class_= "secondline-blog-title")
titles = [title.text for title in title_soup]
titles = [title.replace('\n','').replace('\t','') for title in titles] # not nice, but works
print(len(titles))
titles

21


['Thermal Imagery From Space',
 'I Quit My Job',
 'Monitoring Atmospheric Pollution From Space',
 'Hex Tiles',
 'The Business of Web Maps',
 'Finding Water Leaks From Space',
 'How To Keep Your Satellite Pointing At Earth',
 'Bathymetric Lidar and Blue Carbon',
 'Re-Published – QGIS Offline And In The Field',
 'The Open Geospatial Consortium',
 'Monetizing An Open-Source Geospatial Project',
 'Whitebox Tools Is The Backend To Many Frontends',
 'QGIS Offline And In The Field',
 'Sentinel Hub',
 'Unstructured Data Is Dark Data',
 'What Is Modern GIS?',
 'FOSS4G',
 'Building a web based mapping tool into a business',
 'Digital twins – not just a buzzword',
 'Build Your Own SaaS',
 'Getting Your Dream Job in Earth Observation']

In [29]:
episode_soup = soup.findAll('div', class_= "blog-meta-serie-episode")
episodes = [episode.text for episode in episode_soup]
episodes.insert(16, 'Episode 162')
print(len(episodes))
episodes

21


['Episode 174',
 'Episode 173',
 'Episode 172',
 'Episode 171',
 'Episode 170',
 'Episode 169',
 'Episode 168',
 'Episode 167',
 'Episode 166',
 'Episode 165',
 'Episode 164',
 'Episode 161',
 'Episode 163',
 'Episode 160',
 'Episode 159',
 'Episode 158',
 'Episode 162',
 'Episode 157',
 'Episode 156',
 'Episode 155',
 'Episode 154']

In [31]:
categories_soup = soup.findAll('span', class_= "blog-meta-category-list")
categories = [category.text for category in categories_soup]
categories = [category.replace('\n','').replace('\t','') for category in categories] # not nice, but works
print(len(categories))
categories

21


['Earth Observation, Geospatial Career, Geospatial Concepts, Geospatial Startups, Geospatial Tech and Tools',
 'Geospatial Career, GIS',
 'Earth Observation, Geospatial Tech and Tools',
 'Geospatial Concepts, Geospatial Tech and Tools, GIS',
 'Geospatial Startups, Geospatial Tech and Tools, GIS',
 'Earth Observation, Geospatial Startups, Geospatial Tech and Tools',
 'Earth Observation, Geospatial Concepts, Geospatial Tech and Tools',
 'Earth Observation, Geospatial Concepts, Geospatial Tech and Tools',
 'mapscaping.com',
 'Geospatial Career, Geospatial Startups, Geospatial Tech and Tools',
 'Geospatial Startups, Geospatial Tech and Tools, GIS',
 'Geospatial Tech and Tools, GIS',
 'Geospatial Career, Geospatial Tech and Tools, GIS',
 'Earth Observation, Geospatial Concepts, Geospatial Tech and Tools',
 'Geospatial Startups, Geospatial Tech and Tools, GIS',
 'Geospatial Career, Geospatial Tech and Tools, GIS',
 'Geospatial Tech and Tools',
 'Geospatial Career, Geospatial Startups, Geospa

In [33]:
duration_soup = soup.findAll('span', class_= "blog-meta-time-slt")
durations = [duration.text for duration in duration_soup]
durations.insert(16, '35:53')
print(len(durations))
durations

21


['38:51',
 '13:08',
 '38:53',
 '43:06',
 '52:32',
 '39:50',
 '39:55',
 '44:16',
 '35:52',
 '44:56',
 '59:04',
 '50:35',
 '51:06',
 '41:26',
 '48:00',
 '26:10',
 '35:53',
 '49:35',
 '41:54',
 '44:58',
 '53:12']

In [35]:
df_page3 = pd.DataFrame({'Episode': episodes, 'Date': dates, 'Title': titles, 'Duration': durations, 'Categories': categories})
df_page3['Date'] = pd.to_datetime(df_page3['Date'])
df_page3['Episode'] = df_page3['Episode'].str.replace('Episode ', '')

df_page3.info()
df_page3

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21 entries, 0 to 20
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   Episode     21 non-null     object        
 1   Date        21 non-null     datetime64[ns]
 2   Title       21 non-null     object        
 3   Duration    21 non-null     object        
 4   Categories  21 non-null     object        
dtypes: datetime64[ns](1), object(4)
memory usage: 968.0+ bytes


Episode       Date                                              Title  \
0      174 2022-09-28                         Thermal Imagery From Space   
1      173 2022-09-22                                      I Quit My Job   
2      172 2022-09-21        Monitoring Atmospheric Pollution From Space   
3      171 2022-09-14                                          Hex Tiles   
4      170 2022-09-08                           The Business of Web Maps   
5      169 2022-08-31                     Finding Water Leaks From Space   
6      168 2022-08-18       How To Keep Your Satellite Pointing At Earth   
7      167 2022-08-10                  Bathymetric Lidar and Blue Carbon   
8      166 2022-08-08       Re-Published – QGIS Offline And In The Field   
9      165 2022-08-03                     The Open Geospatial Consortium   
10     164 2022-07-27       Monetizing An Open-Source Geospatial Project   
11     161 2022-07-20    Whitebox Tools Is The Backend To Many Frontends   
12     163 2022-07-20                      QGIS Offline And In The Field   
13     160 2022-07-06                                       Sentinel Hub   
14     159 2022-06-29                     Unstructured Data Is Dark Data   
15     158 2022-06-25                                What Is Modern GIS?   
16     162 2022-06-08                                             FOSS4G   
17     157 2022-06-01  Building a web based mapping tool into a business   
18     156 2022-04-28                Digital twins – not just a buzzword   
19     155 2022-04-20                                Build Your Own SaaS   
20     154 2022-04-13        Getting Your Dream Job in Earth Observation   

   Duration                                         Categories  
0     38:51  Earth Observation, Geospatial Career, Geospati...  
1     13:08                             Geospatial Career, GIS  
2     38:53       Earth Observation, Geospatial Tech and Tools  
3     43:06  Geospatial Concepts, Geospatial Tech and Tools...  
4     52:32  Geospatial Startups, Geospatial Tech and Tools...  
5     39:50  Earth Observation, Geospatial Startups, Geospa...  
6     39:55  Earth Observation, Geospatial Concepts, Geospa...  
7     44:16  Earth Observation, Geospatial Concepts, Geospa...  
8     35:52                                     mapscaping.com  
9     44:56  Geospatial Career, Geospatial Startups, Geospa...  
10    59:04  Geospatial Startups, Geospatial Tech and Tools...  
11    50:35                     Geospatial Tech and Tools, GIS  
12    51:06  Geospatial Career, Geospatial Tech and Tools, GIS  
13    41:26  Earth Observation, Geospatial Concepts, Geospa...  
14    48:00  Geospatial Startups, Geospatial Tech and Tools...  
15    26:10  Geospatial Career, Geospatial Tech and Tools, GIS  
16    35:53                          Geospatial Tech and Tools  
17    49:35  Geospatial Career, Geospatial Startups, Geospa...  
18    41:54  Geospatial Concepts, Geospatial Tech and Tools...  
19    44:58  Earth Observation, Geospatial Startups, Geospa...  
20    53:12          Earth Observation, Geospatial Career, GIS

In [37]:
df_complete = pd.concat([result_df, df_page3], ignore_index = True)

In [39]:
final_df = df_complete.sort_values(by='Date')
final_df

Episode       Date                                              Title  \
215       1 2019-01-14  The future of collecting and updating geospati...   
214       2 2019-02-08  Indoor mapping and navigation: Manage, visuali...   
213       3 2019-03-10                    Bellerby & Co – The globemakers   
212       4 2019-04-02         freelance mappers create maps for machines   
211       5 2019-04-08  Powering location intelligence with geo social...   
210       6 2019-04-16              Data discovery – the way it should be   
209       7 2019-04-25  Augmented reality will change the way you thin...   
208       8 2019-05-03  Geo-tagged audio – another way of augmenting r...   
207       9 2019-05-08  Proof of location, bringing the blockchain to ...   
206      10 2019-05-15                Mapping Personalised Workplace Risk   
205      11 2019-05-22  Translating between machine and human when tal...   
204      12 2019-05-29           Tracking global air traffic in real time   
203      13 2019-06-05     A mapping platform for your outdoor adventures   
202      14 2019-06-12  Scaling map data generation using computer vision   
201      15 2019-06-19  A diary of your location – recording your trav...   
200      16 2019-06-25  Positioning as a service and the role of smart...   
199      17 2019-07-03             Creating Global Activity Specific Maps   
198      18 2019-07-10  Mapping the way we relate to the urban environ...   
197      19 2019-07-17           A geospatial story and some housekeeping   
196      20 2019-07-31  The Google Drive of geospatial – curation over...   
195      21 2019-08-08  Immersive maps give context to data – this is ...   
194      22 2019-08-14  Highways in the sky, risk islands and drone fl...   
193      23 2019-08-21  Crowd source clean the planet – a geospatial a...   
192      24 2019-08-28    Geospatial support for humanitarian emergencies   
191      25 2019-09-04            Inspirational geospatial visualizations   
190      26 2019-09-11                   Geospatial gives context to risk   
189      27 2019-09-18                      The location context platform   
188      28 2019-09-25  Using the geomagnetic field of buildings to na...   
187      29 2019-10-02  Geospatial visualisations matter – Civil engin...   
186      30 2019-10-09    A marketplace for geospatial data and workflows   
185      31 2019-10-16     Remote sensing and future of earth observation   
184      32 2019-10-23                       Building footprints at scale   
183      33 2019-10-30  Geomob – Geoinnovation for the people by the p...   
182      34 2019-11-06                      Creating maps for journalists   
181      35 2019-11-13  Mapping the interface between vehicle traffic ...   
180      36 2019-11-20                     Building geospatial truth sets   
179      37 2019-11-27             Don’t just make the dots, connect them   
178      38 2019-12-04  Addressing the world – a coordinate system fro...   
177      39 2019-12-11          Remote sensing – building the globe in 3D   
176      40 2019-12-18  Travel Time – A better way of doing location b...   
175      41 2019-12-26  Trends in geospatial and initiatives in the co...   
174      42 2020-01-15  The evolution of geospatial file formats and d...   
173      43 2020-01-22  Geospatial intelligence from the sky, automati...   
172      44 2020-01-29              Is R the language of geospatial data?   
171      45 2020-02-05  Geo-coding as a tool for risk assessment and t...   
170      46 2020-02-12  PostgreSQL – An open source geospatial databas...   
169      47 2020-02-19  Earth Observation – machine learning and artif...   
168      48 2020-02-26                              Serverless Geospatial   
167      49 2020-03-03                                A Case Study in GIS   
166      50 2020-03-12            QGIS – An opensource geospatial project   
165      51 2020-03-18  Disrupting ESRI – Why it is hard to build a bi...   


In [106]:
final_df.to_csv('shiny-map/data/MapScaping_scraped.tsv', sep='\t', index = False)

### add interviewee info

this is the more tricky part. Daniel does not list the names or contact information of his interview partners is a structured way on his website. Therefore, I have to manually go through the transcripts (those he has, luckily!) to find the names of the interview partners and then browse the internet to find social media profiles etc. with data on them

#### Variables I try to find for each interviewee are:
* (interviewee id)
* first name
* last name
* place of residence
* gender (based on pronouns on LinkedIn or how other people/websites refer to them)
* position/seniority *at time of interview*
* weblinks (to LinkedIn/Twitter/personal page and to company page)
* company name *at time of interview*

#### Shortcomings
* not everyone is 'from' somewhere. Several interviewees moved quite often in their lives and don't consider themselves to be from a specific place
* gender is assumed from name & photos if not explicitly stated on LinkedIn or clear from how Daniel refers to them
* the location I use is point data, while some interviewees state a region (or even country) as their location
* quite a few companies that the interviewees worked at the time of the interview were acquired by other firms or were closed
* 

some names can't be written correctly because of UTF8 encoding, e.g.
* Lyden Foust
* Josh Kopecek
* Markus Müller

* several place names (especially PL, CZ) 

Not sure whether the map is too much advertisement for the interview partners / their companies. Daniel does not want the podcast to be a sales pitch for them.

---

I use Excel to type in the data that I found online.
Create custom function to read csv as pandas function causes troubles because of UTF-8 encoing issues

In [102]:
import os

## from 
## https://stackoverflow.com/questions/48812580/python-pandas-unicodedecodeerror-utf-8-codec-cant-decode-byte-0xcd-in-pos

def read_csv(filepath):
     if os.path.splitext(filepath)[1] != '.csv':
          return  # or whatever
     seps = [',', ';', '\t']                    # ',' is default
     encodings = [None, 'utf-8', 'ISO-8859-1']  # None is default
     for sep in seps:
         for encoding in encodings:
              try:
                  return pd.read_csv(filepath, encoding=encoding, sep=sep)
              except Exception:  # should really be more specific 
                  pass
     raise ValueError("{!r} is has no encoding in {} or seperator in {}"
                      .format(filepath, encodings, seps))

In [104]:
df = read_csv("shiny-map/data/MapScaping_extended.csv")

In [65]:
# double square brackets to keep it as DataFrame (instead of Series)
locations = df[['location']]
# drop duplicate locations to make geocoding faster and avoid merge complications later on
locations = locations.drop_duplicates(subset='location')

locations #.info()

location
0                               San Francisco Bay Area
1                                                  NaN
2    London, England, Islington, Borough of Hackney...
4                                Cincinnati, Ohio, USA
5                      Auckland, Auckland, New Zealand
6                             Toronto, Ontario, Canada
7                  Okehampton, England, United Kingdom
8                           San Diego, California, USA
9                                   Austin, Texas, USA
10           Heilbad Heiligenstadt, Thuringia, Germany
11                   Chicago, United States of America
12                             Genf, Genf, Switzerland
13                            Malmo, Skane Lan, Sweden
14                                 London, England, UK
16                                          London, UK
17                                  Washington DC, USA
18                          Vejle, Syddanmark, Denmark
19                          Stanford, Kalifornien, USA
20                     Wimberley, Texas, United States
21                              Detroit, Michigan, USA
22                     Chapel Hill, Nord-Carolina, USA
23                                      Bolzano, Italy
24                                      Manchester, UK
26                             New York, New York, USA
27                          Helsinki, Uusimaa, Finland
28                Kitchener-Cambridge-Waterloo, Canada
29                             Berlin, Berlin, Germany
30                                         Bristol, UK
31                     Albany, New York, United States
33               Eindhoven, North Brabant, Netherlands
34                   New York, New York, United States
36                   St Paul, Minnesota, United States
37                     London, England, United Kingdom
38                Trumbull, Connecticut, United States
39                  Islington, England, United Kingdom
41                    Surrey, British Columbia, Canada
43                         Nuremberg, Bavaria, Germany
44                                      Grasse, France
45             Fremantle, Western Australia, Australia
46            Charlottesville, Virginia, United States
47                           Yorkshire, United Kingdom
49                        Albuquerque, New Mexico Area
52                  Victoria, British Columbia, Canada
54                           Colorado Springs, CO, USA
55              Albuquerque, New Mexico, United States
57            Charlotte, North Carolina, United States
59                             Vienna, Vienna, Austria
60            San Francisco, California, United States
61                            Calgary, Alberta, Canada
63                                       Isen, Germany
64                    Calumet, Michigan, United States
66                    Newport, England, United Kingdom
67                 Vancouver, British Columbia, Canada
69                  Seattle, Washington, United States
73                 Stanford, California, United States
78                                Grand Ledge, MI, USA
79                 Winchester, England, United Kingdom
80                                        South Africa
81                           Waterloo, Ontario, Canada
82                            Toronto, Ontario, Kanada
84                            Vevey, Vaud, Switzerland
85               Bonn, North Rhine-Westphalia, Germany
88                              Greater Melbourne Area
90                    Norwich, England, United Kingdom
91     Washington, District of Columbia, United States
92                              Prague, Czech Republic
93                                      Greater Boston
96                   Irvine, California, United States
97               Fort Collins, Colorado, United States
98                   Sidmouth, England, United Kingdom
99                 Knoxville, Tennessee, United States
101                       Salem, Oregon, United States
102                            

### geocode addresses

use Nomatim API for this

In [67]:
import pandas as pd
import requests
import time

# Load addresses from CSV
addresses_df = locations # pd.read_csv('location.csv')
addresses = addresses_df['location'].tolist()

# Function to geocode address using Nominatim API
def geocode_address(address):
    url = 'https://nominatim.openstreetmap.org/search'
    headers = {'User-Agent': 'Nicolas'}
    params = {'q': address, 'format': 'json'}
    
    response = requests.get(url, headers=headers, params=params)
    if response.status_code == 200:
        results = response.json()
        if results:
            return results[0]['lat'], results[0]['lon']  # Return the latitude and longitude of the first result
    return None, None  # Return None if no results or an error occurred

# Create a list to hold geocoded results
geocoded_addresses = []

# Geocode each address
for address in tqdm(addresses):
    lat, lon = geocode_address(address)
    geocoded_addresses.append({'address': address, 'latitude': lat, 'longitude': lon})
    time.sleep(1)  # Sleep to respect Nominatim's usage policy

# Convert results to a DataFrame
geocoded_df = pd.DataFrame(geocoded_addresses)

# Optionally, save the geocoded results to a new CSV file
geocoded_df.to_csv('shiny-map/data/geocoded_addresses.csv', index=False)

print('Geocoding complete. Results saved to geocoded_addresses.csv.')


100%|██████████████████████████████████████████████████████████████████████████████| 144/144 [04:04<00:00,  1.70s/it]

Geocoding complete. Results saved to geocoded_addresses.csv.


In [69]:
geocoded_df
# manually insert coordinates for...
# ...Bellerby & Co., as it's too specific for nomatim 
geocoded_df.at[2, 'latitude'] = 51.5625709
geocoded_df.at[2, 'longitude'] = -0.0788484
# ...Cologne-Bonn-Region in Germany, as it's unknown to nomatim 
geocoded_df.at[112, 'latitude'] = 50.8285133
geocoded_df.at[112, 'longitude'] = 6.9960294
# ...NaN, as Nominatims geocoding obviously makes no sense at all :D
geocoded_df.at[1, 'latitude'] = 'NaN'
geocoded_df.at[1, 'longitude'] = 'NaN'

geocoded_df

address             latitude  \
0                               San Francisco Bay Area           37.7884969   
1                                                  NaN                  NaN   
2    London, England, Islington, Borough of Hackney...            51.562571   
3                                Cincinnati, Ohio, USA           39.1014537   
4                      Auckland, Auckland, New Zealand           -36.852095   
5                             Toronto, Ontario, Canada           43.6534817   
6                  Okehampton, England, United Kingdom           50.7376667   
7                           San Diego, California, USA           32.7174202   
8                                   Austin, Texas, USA           30.2711286   
9            Heilbad Heiligenstadt, Thuringia, Germany           51.3756186   
10                   Chicago, United States of America           41.8755616   
11                             Genf, Genf, Switzerland           46.2017559   
12                            Malmo, Skane Lan, Sweden           55.6052931   
13                                 London, England, UK           51.5074456   
14                                          London, UK           51.5074456   
15                                  Washington DC, USA           38.8950368   
16                          Vejle, Syddanmark, Denmark           55.7066917   
17                          Stanford, Kalifornien, USA   37.426540700000004   
18                     Wimberley, Texas, United States           29.9976807   
19                              Detroit, Michigan, USA           42.3315509   
20                     Chapel Hill, Nord-Carolina, USA           35.9131542   
21                                      Bolzano, Italy   46.655945450000004   
22                                      Manchester, UK           53.4794892   
23                             New York, New York, USA           40.7127281   
24                          Helsinki, Uusimaa, Finland           60.1674881   
25                Kitchener-Cambridge-Waterloo, Canada            43.415134   
26                             Berlin, Berlin, Germany           52.5103817   
27                                         Bristol, UK           51.4538022   
28                     Albany, New York, United States           42.6511674   
29               Eindhoven, North Brabant, Netherlands           51.4392648   
30                   New York, New York, United States           40.7127281   
31                   St Paul, Minnesota, United States           44.9497487   
32                     London, England, United Kingdom           51.5074456   
33                Trumbull, Connecticut, United States           41.2428742   
34                  Islington, England, United Kingdom           51.5384287   
35                    Surrey, British Columbia, Canada           49.1913033   
36                         Nuremberg, Bavaria, Germany            49.453872   
37                                      Grasse, France           43.6589011   
38             Fremantle, Western Australia, Australia          -32.0534086   
39            Charlottesville, Virginia, United States            38.029306   
40                           Yorkshire, United Kingdom           53.9825271   
41                        Albuquerque, New Mexico Area           35.1357749   
42                  Victoria, British Columbia, Canada           48.4283182   
43                           Colorado Springs, CO, USA           38.8339578   
44              Albuquerque, New Mexico, United States           35.0841034   
45            Charlotte, North Carolina, United States           35.2272086   
46                             Vienna, Vienna, Austria           48.1857192   
47            San Francisco, California, United States           37.7792588   
48                            Calgary, Alberta, Canada           51.0456064   
49                                       Isen, Germany            48.274057   
50                

In [71]:
# drop rows without location / coordinates
# create explicit copy of the DataFrame
geocoded_df_clean = geocoded_df.dropna().copy()
geocoded_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 143 entries, 0 to 143
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   address    143 non-null    object
 1   latitude   143 non-null    object
 2   longitude  143 non-null    object
dtypes: object(3)
memory usage: 4.5+ KB


In [73]:
# transform into float and round to 7 decimal points (approx 1cm)
geocoded_df_clean.loc[:, 'latitude'] = geocoded_df_clean['latitude'].astype(float).round(7)
geocoded_df_clean.loc[:, 'longitude'] = geocoded_df_clean['longitude'].astype(float).round(7)

In [75]:
geocoded_df_clean = geocoded_df_clean.rename(columns={'address': 'location'})
geocoded_df_clean.info()
df#.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 143 entries, 0 to 143
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   location   143 non-null    object 
 1   latitude   143 non-null    float64
 2   longitude  143 non-null    float64
dtypes: float64(2), object(1)
memory usage: 4.5+ KB


Episode      Date                                              Title  \
0          1  14.01.19  The future of collecting and updating geospati...   
1          2  08.02.19  Indoor mapping and navigation: Manage, visuali...   
2          3  10.03.19                    Bellerby & Co Ð The globemakers   
3          4  02.04.19         freelance mappers create maps for machines   
4          5  08.04.19  Powering location intelligence with geo social...   
5          6  16.04.19              Data discovery Ð the way it should be   
6          7  25.04.19  Augmented reality will change the way you thin...   
7          8  03.05.19  Geo-tagged audio Ð another way of augmenting r...   
8          9  08.05.19  Proof of location, bringing the blockchain to ...   
9         10  15.05.19                Mapping Personalised Workplace Risk   
10        11  22.05.19  Translating between machine and human when tal...   
11        12  29.05.19           Tracking global air traffic in real time   
12        13  05.06.19     A mapping platform for your outdoor adventures   
13        14  12.06.19  Scaling map data generation using computer vision   
14        15  19.06.19  A diary of your location Ð recording your trav...   
15        16  25.06.19  Positioning as a service and the role of smart...   
16        17  03.07.19             Creating Global Activity Specific Maps   
17        18  10.07.19  Mapping the way we relate to the urban environ...   
18        19  17.07.19           A geospatial story and some housekeeping   
19        20  31.07.19  The Google Drive of geospatial Ð curation over...   
20        21  08.08.19  Immersive maps give context to data Ð this is ...   
21        22  14.08.19  Highways in the sky, risk islands and drone fl...   
22        23  21.08.19  Crowd source clean the planet Ð a geospatial a...   
23        24  28.08.19    Geospatial support for humanitarian emergencies   
24        25  04.09.19            Inspirational geospatial visualizations   
25        26  11.09.19                   Geospatial gives context to risk   
26        27  18.09.19                      The location context platform   
27        28  25.09.19  Using the geomagnetic field of buildings to na...   
28        29  02.10.19  Geospatial visualisations matter Ð Civil engin...   
29        30  09.10.19    A marketplace for geospatial data and workflows   
30        31  16.10.19     Remote sensing and future of earth observation   
31        32  23.10.19                       Building footprints at scale   
32        33  30.10.19  Geomob Ð Geoinnovation for the people by the p...   
33        34  06.11.19                      Creating maps for journalists   
34        35  13.11.19  Mapping the interface between vehicle traffic ...   
35        36  20.11.19                     Building geospatial truth sets   
36        37  27.11.19             DonÕt just make the dots, connect them   
37        38  04.12.19  Addressing the world Ð a coordinate system fro...   
38        39  11.12.19          Remote sensing Ð building the globe in 3D   
39        40  18.12.19  Travel Time Ð A better way of doing location b...   
40        41  26.12.19  Trends in geospatial and initiatives in the co...   
41        42  15.01.20  The evolution of geospatial file formats and d...   
42        43  22.01.20  Geospatial intelligence from the sky, automati...   
43        44  29.01.20              Is R the language of geospatial data?   
44        45  05.02.20  Geo-coding as a tool for risk assessment and t...   
45        46  12.02.20  PostgreSQL Ð An open source geospatial databas...   
46        47  19.02.20  Earth Observation Ð machine learning and artif...   
47        48  26.02.20                              Serverless Geospatial   
48        49  03.03.20                                A Case Study in GIS   
49        50  12.03.20            QGIS Ð An opensource geospatial project   
50        51  18.03.20  Disrupting ESRI Ð Why it is hard to build a bi...   
5

In [77]:
# merge geocoded coordinates to main dataframe
result_df = pd.merge(df, geocoded_df_clean, on='location', how='left')

### feature engineering

In [79]:
result_df['Date'] = pd.to_datetime(result_df['Date'], format='%d.%m.%y')
# check that datetime conversion worked properly
result_df = result_df.sort_values(by=['Date'])
result_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 238 entries, 0 to 237
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Episode           238 non-null    int64         
 1   Date              238 non-null    datetime64[ns]
 2   Title             238 non-null    object        
 3   Duration          238 non-null    object        
 4   Categories        201 non-null    object        
 5   interviewee_id    226 non-null    float64       
 6   first_name        237 non-null    object        
 7   last_name         236 non-null    object        
 8   gender            237 non-null    object        
 9   position          235 non-null    object        
 10  interviewee_link  236 non-null    object        
 11  company_name      235 non-null    object        
 12  location          235 non-null    object        
 13  page_on_website   238 non-null    int64         
 14  company_link      236 non-

In [83]:
# Creating a combined name column for convenience
result_df['full_name'] = result_df['first_name'] + ' ' + result_df['last_name']

# Counting occurrences and creating a new column
result_df['number_of_interviews'] = result_df.groupby('full_name')['full_name'].transform('count') #.astype(int)

In [85]:
# who appears how often?
result_df.sort_values(by=['number_of_interviews', 'full_name'], ascending = False)

Episode       Date                                              Title  \
18        19 2019-07-17           A geospatial story and some housekeeping   
40        41 2019-12-26  Trends in geospatial and initiatives in the co...   
86        87 2020-12-03                                   My Story, my why   
126      127 2021-09-22                            Geospatial Side Hustles   
139      140 2021-12-22                                      Voice Message   
140      141 2022-01-05                                 What Is Your Idea?   
146      147 2022-02-16               Business Ideas For Geospatial People   
172      173 2022-09-22                                      I Quit My Job   
52        53 2020-04-01    Dynamic vector tiles straight from the database   
76        77 2020-09-16                  Spatial SQL Ð GIS without the GIS   
106      107 2021-05-06         GDAL Ð Geospatial Data Abstraction Library   
198      199 2023-05-17                             Rasters In A Database?   
204      205 2023-06-28  PostgreSQL Ð Listen and Notify Clients In Real...   
49        50 2020-03-12            QGIS Ð An opensource geospatial project   
58        59 2020-05-13  Two mobile data collection apps you need to kn...   
74        75 2020-09-02   Self-employment in the GIS / Geospatial industry   
108      109 2021-05-19         An Introduction To Artificial Intelligence   
176      177 2022-10-28                            Artificial Intelligence   
197      198 2023-05-10                  ChatGPT and Large Language Models   
30        31 2019-10-16     Remote sensing and future of earth observation   
51        52 2020-03-25              Earth Observation, platforms and data   
77        78 2020-09-23  Being self employed in the earth observation s...   
98        99 2021-03-10  Getting Where You Want To Go In Your Geospatia...   
175      176 2022-10-19                                  Mid-Career Change   
93        94 2021-02-04  Skills, Leadership, Mentorship and the Geospat...   
212      213 2023-10-04                        Strategic Buy-In For FOSS4G   
111      112 2021-06-09                                          Geoserver   
116      117 2021-07-14                                           MapStore   
99       100 2021-03-17                    Introducing Google Earth Engine   
220      221 2024-01-09                                             GeeMap   
161      162 2022-07-20                      QGIS Offline And In The Field   
165      166 2022-08-08       Re-Published Ð QGIS Offline And In The Field   
100      101 2021-03-24  Open Geospatial Standards Ð shared standards t...   
164      165 2022-08-03                     The Open Geospatial Consortium   
36        37 2019-11-27             DonÕt just make the dots, connect them   
48        49 2020-03-03                                A Case Study in GIS   
132      133 2021-11-03  Collecting And Processing Aerial Imagery At Scale   
184      185 2023-02-01               Aerial Imagery: The State Of The Art   
44        45 2020-02-05  Geo-coding as a tool for risk assessment and t...   
75        76 2020-09-11                                     Elastic Search   
168      169 2022-08-31                     Finding Water Leaks From Space   
230      231 2024-06-05                   Why You Should Care About L Band   
160      161 2022-07-20    Whitebox Tools Is The Backend To Many Frontends   
163      164 2022-07-27       Monetizing An Open-Source Geospatial Project   
120      121 2021-08-12  The Problem With Satellite Data Is ÉThat It Is...   
121      122 2021-08-19  Building A Personal Brand In The Geospatial In...   
101      102 2021-03-31        OpenStreetMap is a community of communities   
210      211 2023-08-30        Overture Maps And The Daylight Distribution   
96        97 2021-02-24     Introduction to Synthetic-aperture Radar (SAR)   
153      154 2022-04-13        Getting Your Dream Job in Earth Observation   
128      129 2021-10-06    

In [87]:
result_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 238 entries, 0 to 237
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Episode               238 non-null    int64         
 1   Date                  238 non-null    datetime64[ns]
 2   Title                 238 non-null    object        
 3   Duration              238 non-null    object        
 4   Categories            201 non-null    object        
 5   interviewee_id        226 non-null    float64       
 6   first_name            237 non-null    object        
 7   last_name             236 non-null    object        
 8   gender                237 non-null    object        
 9   position              235 non-null    object        
 10  interviewee_link      236 non-null    object        
 11  company_name          235 non-null    object        
 12  location              235 non-null    object        
 13  page_on_website     

In [89]:
# subset only columns that are relevant
subset = result_df[['Episode', 'Date', 'Title', 'Duration', 'Categories', 'first_name', 'last_name', 'interviewee_link', 'gender', 'position', 'company_name', 'company_link', 'location', 'latitude', 'longitude', 'full_name', 'number_of_interviews']]

In [91]:
subset.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 238 entries, 0 to 237
Data columns (total 17 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Episode               238 non-null    int64         
 1   Date                  238 non-null    datetime64[ns]
 2   Title                 238 non-null    object        
 3   Duration              238 non-null    object        
 4   Categories            201 non-null    object        
 5   first_name            237 non-null    object        
 6   last_name             236 non-null    object        
 7   interviewee_link      236 non-null    object        
 8   gender                237 non-null    object        
 9   position              235 non-null    object        
 10  company_name          235 non-null    object        
 11  company_link          236 non-null    object        
 12  location              235 non-null    object        
 13  latitude            

In [108]:
# save as tab-separated file so that commas in addresses don't mess it up
subset.to_csv('shiny-map/data/geocoded_addresses.tsv', sep='\t', index = False)

### future tasks

### To Do Preprocessing
* rename variables to be meaningful
* create variable with count of interviews per person
* change data types to time / timedelta? Or only later in postgis??





### To Do in Angular Frontend
* jitter points so that they are not exactly on top of each other
* clustering of points that overlap too much (solve css problems)
    * use colorcoding for percentage of male/female interviewees in cluster
* handle special characters correctly (in names)
* include more in popup
    * interviewee name (make link clickable)
    * episode
    * company name (make link clickable)
* allow filtering for 
    * category (use regex to check whether category is in string)
    * gender
    * date (range filter?)
    * number of interviews (checkbox)
    * interviewee name?
    * company name?

To introduce a jitter of approximately 1 kilometer to the locations in your interviewees table to ensure that identical locations don’t overlay each other, you can use the ST_Translate function in PostGIS. This function shifts the geometry by a specified amount in the x (longitude) and y (latitude) directions. The amount to shift can be determined based on the degree equivalent of the desired distance at the given latitude (keeping in mind that a degree of longitude varies in actual distance depending on the latitude, but for small distances like 1km, this can be approximated fairly accurately).

For the jitter, we can use random offsets in both the latitude and longitude directions. Given that 1 degree of latitude is approximately 111 kilometers, a shift of about 0.009° in any direction would approximate to 1 kilometer. For longitude, this value needs to be adjusted based on the latitude due to the varying distance covered by a degree of longitude, but for simplicity, we'll use the same approximation, which is reasonable for small distances and near the equator.

This SQL statement will move each point in a random direction by up to approximately 1 kilometer. The RANDOM() function generates a value between 0 and 1, so RANDOM() * 0.018 - 0.009 will produce a shift ranging from -0.009° to +0.009° in both latitude and longitude, effectively creating a jitter around the original point.

not needed
* interviewee_id
* page_on_website
* incomplete

**Attention!!!** 
* there are two different Eric Jensen's
* Peter Petrik's episode was re-published because of audio-quality